internal/component/pyroscope/relabel/relabel.go

```diff
@@ -0,0 +1,333 @@
+
+	c.opts.OnStateChange(Exports{
+		Receiver: c,
+		Rules:    newArgs.RelabelConfigs,
+	})
+
+	return nil
+}
+
+func (c *Component) Append(ctx context.Context, lbls labels.Labels, samples []*pyroscope.RawSample) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	newLabels, keep, err := c.relabel(lbls)
+	if err != nil {
+		return err
+	}
+	if !keep {
+		level.Debug(c.opts.Logger).Log("msg", "profile dropped by relabel rules", "labels", lbls.String())
+		return nil
+	}
+
+	return c.fanout.Appender().Append(ctx, newLabels, samples)
+}
+
+func (c *Component) AppendIngest(ctx context.Context, profile *pyroscope.IncomingProfile) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	if profile.Labels.IsEmpty() {
+		return c.fanout.Appender().AppendIngest(ctx, profile)
+	}
+
+	newLabels, keep, err := c.relabel(profile.Labels)
+	if err != nil {
+		return fmt.Errorf("processing labels: %w", err)
+	}
+	if !keep {
+		c.metrics.profilesDropped.Inc()
+		level.Debug(c.opts.Logger).Log("msg", "profile dropped by relabel rules")
+		return nil
+	}
+
+	profile.Labels = newLabels
+	return c.fanout.Appender().AppendIngest(ctx, profile)
+}
+
+func (c *Component) Appender() pyroscope.Appender {
+	return c
+}
+
+type cacheItem struct {
+	original  model.LabelSet
+	relabeled model.LabelSet
+}
+
+// relabel applies the configured relabeling rules to the input labels
+// Returns the new labels, whether to keep the profile, and any error
+func (c *Component) relabel(lbls labels.Labels) (labels.Labels, bool, error) {
+	labelSet := toModelLabelSet(lbls)
+	hash := labelSet.Fingerprint()
+
+	// Check cache
+	if result, keep, found := c.getCacheEntry(hash, labelSet); found {
+		return result, keep, nil
+	}
+
+	// Apply relabeling
+	// builder := labels.NewBuilder(lbls)
+	builder := labels.NewBuilder(lbls)
+	keep := relabel.ProcessBuilder(builder, c.rcs...)
+	if !keep {
+		c.metrics.profilesDropped.Inc()
+		return labels.Labels{}, false, nil
+	}
+
+	newLabels := builder.Labels()
+
+	// Cache result
+	c.addToCache(hash, labelSet, newLabels)
+
+	return newLabels, true, nil
+}
+
+func (c *Component) getCacheEntry(hash model.Fingerprint, labelSet model.LabelSet) (labels.Labels, bool, bool) {
+	if val, ok := c.cache.Get(hash); ok {
+		for _, item := range val.([]cacheItem) {
+			if labelSet.Equal(item.original) {
+				c.metrics.cacheHits.Inc()
+				if len(item.relabeled) == 0 {
+					c.metrics.profilesDropped.Inc()
+					return labels.Labels{}, false, true
+				}
+				return toLabelsLabels(item.relabeled), true, true
+			}
+		}
+	}
+	c.metrics.cacheMisses.Inc()
+	return labels.Labels{}, false, false
+}
+
+func (c *Component) addToCache(hash model.Fingerprint, original model.LabelSet, relabeled labels.Labels) {
+	var cacheValue []cacheItem
+	if val, exists := c.cache.Get(hash); exists {
+		cacheValue = val.([]cacheItem)
```

Comment:
```suggestion
		cacheValue = val.([]cacheItem)
```

Nit: lruv2 has generics and is what prometheus.relabel uses